### PR TITLE
RoboCup 2019 world update changes

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description
Changes from RoboCup 2019. Slight changes to how various parts of the world and gamestate are updated.

### Testing Done
unit tests still pass, validated with RoboCup gameplay

### Resolved Issues
None

### Length Justification
None

### CODE REVIEW

File: src/thunderbots/software/ai/world/team.cpp [hunk 1]
```diff
@@ -83,8 +83,9 @@ void Team::removeExpiredRobots(const Timestamp& timestamp)
         Duration time_diff = timestamp - it->lastUpdateTimestamp();
         if (time_diff.getSeconds() < 0)
         {
-            throw std::invalid_argument(
-                "Error: tried to remove a robot at a negative time");
+            LOG(WARNING) << "Error: tried to remove a robot at a negative time"
+                         << std::endl;
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/team.cpp]: We don't need to use an `endl` with `LOG`

File: src/thunderbots/software/ai/world/world.cpp [hunk 2]
```diff
@@ -23,7 +23,10 @@ World::World(const Field &field, const Ball &ball, const Team &friendly_team,
       ball_(ball),
       friendly_team_(friendly_team),
       enemy_team_(enemy_team),
-      game_state_()
+      game_state_(),
+      // Store a small buffer of previous refbox game states so we can filter out noise
+      // Do not make this value smaller than 3
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/world.cpp]: See my comment below, I think we can remove this restriction

File: src/thunderbots/software/ai/world/world.cpp [hunk 3]
```diff
@@ -115,7 +118,18 @@ Team &World::mutableEnemyTeam()
 
 void World::updateRefboxGameState(const RefboxGameState &game_state)
 {
-    game_state_.updateRefboxGameState(game_state, ball_);
+    refbox_game_state_history.push_back(game_state);
+    // Take the consensus of the past 3 refbox messages
+    if (refbox_game_state_history.size() >= 3 &&
+        refbox_game_state_history[0] == refbox_game_state_history[1] &&
+        refbox_game_state_history[0] == refbox_game_state_history[2])
+    {
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/world.cpp:127]: Why not take the consensus of the last `n` refbox messages, where `n` is the array length?

File: src/thunderbots/software/ai/world/world.cpp [hunk 4]
```diff
@@ -115,7 +118,18 @@ Team &World::mutableEnemyTeam()
...
+        game_state_.updateRefboxGameState(game_state, ball_);
+    }
+    else
+    {
+        game_state_.updateRefboxGameState(game_state_.getRefboxGameState(), ball_);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/world.cpp]: This really feels like we should have `updateRefboxGameState(RefboxGameState)` and `updateBallState(Ball)`

File: src/thunderbots/software/ai/world/world.cpp [hunk 5]
```diff
@@ -32,25 +35,25 @@ World::World(const Field &field, const Ball &ball, const Team &friendly_team,
...
 {
-    friendly_team_.updateState(new_friendly_team_data);
+    friendly_team_ = new_friendly_team_data;
     updateTimestamp(getMostRecentTimestampFromMembers());
 }
 
 void World::updateEnemyTeamState(const Team &new_enemy_team_data)
 {
-    enemy_team_.updateState(new_enemy_team_data);
+    enemy_team_ = new_enemy_team_data;
```
Comment (garethellis0) [inline src/thunderbots/software/ai/world/world.cpp]: What was wrong with what we were doing before? Should we remove the `updateState` methods?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/world/world.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/world.cpp]: It was a mix of making most update functions function like assignment (which was being done in `ai/main.cpp` before) but let some update state. It was a change necessary to act like we're "copying" the world data into the `ai_logic` node, and only update the `gamestate` state.

I think we should revisit the usage of the update functions when we create the sensor fusion abstraction. I'd prefer to leave these as they are for now unless we have a good reason to change them immediately.

File: src/thunderbots/software/ai/world/world.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/world.cpp]: You mean updating the ball used in the "gamestate statemaching" separately? Right now the ball and gamestate are updated separately, this code is not updating the ball.

File: src/thunderbots/software/ai/world/world.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/world.cpp]: done

File: src/thunderbots/software/ai/world/world.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/world.cpp:127]: done

File: src/thunderbots/software/ai/world/world.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/world/world.cpp]: I'm still confused what the issue is with calling `updateState` on our member variables here?

File: src/thunderbots/software/ai/world/world.cpp (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/world/world.cpp]: What is this code doing if not updating the ball in `game_state_`?

File: src/thunderbots/software/ai/world/team.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/team.cpp]: fixed

**Comment by MathewMacDougall** [comment]
@garethellis0 	ready for re-review

File: src/thunderbots/software/ai/world/world.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/world/world.cpp]: After some discussion offline and some thought, I don't think we should make this change. I have what is essentially a fix for this issue in an upcoming PR, and this obfuscates the concept of "updating" something.

File: src/thunderbots/software/ai/world/world.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/world/world.cpp]: fixed

**Comment by MathewMacDougall** [comment]
@garethellis0 	ready for re-review

**Review by garethellis0 — APPROVED** [review]
LGTM 👍 🎉

**Pull request merged**